elastic-apm-agent/src/main/java/co/elastic/apm/agent/premain/ShadedClassLoader.java

```diff
@@ -68,9 +74,58 @@ public ShadedClassLoader(File jar, ClassLoader parent, String customPrefix) thro
         }
     }
 
+    /**
+     * A child-first implementation for class loading by searching for classes in the following order:
+     *
+     * <ol>
+     *   <li><p> Invoke {@link #findLoadedClass(String)} to check if the class
+     *   has already been loaded.  </p></li>
+     *
+     *   <li><p> Invoke {@link #findClass(String)} to find the
+     *   class locally. The way to guarantee only local lookup is by searching only for the
+     *   shaded form. </p></li>
+     *
+     *   <li><p> If not found so far, invoke the {@link ClassLoader#loadClass(String) loadClass}
+     *   method. This will start a parent-first lookup, eventually looking locally through
+     *   {@link ClassLoader#findClass(String)}, which will only do the normal {@link URLClassLoader#ucp}
+     *   lookup, which can only find a limited number of classes used for the agent very early initialization.
+     *   Note that this would also cause a redundant additional {@link ClassLoader#getResource(String)}
+     *   lookup, but we guard from that through a {@link ThreadLocal}. </p></li>
+     * </ol>
+     *
+     * @param  name
+     *         The <a href="#binary-name">binary name</a> of the class
+     *
+     * @param  resolve
+     *         If {@code true} then resolve the class
+     *
+     * @return  The resulting {@code Class} object
+     *
+     * @throws  ClassNotFoundException
+     *          If the class could not be found
+     */
+    @Override
+    protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        synchronized (getClassLoadingLock(name)) {
+            try {
+                // First, check if the class has already been loaded
+                Class<?> c = findLoadedClass(name);
+                if (c == null) {
+                    c = findClass(name);
+                    if (resolve) {
+                        resolveClass(c);
+                    }
+                }
```

Comment:
In the case where the class name does not start with `customPrefix`, could we avoid attempting the local lookup and directly call the parent instead ?

Reply:
Class name should never start with the custom prefix, as there is no `agent.` package in reality.
Note that everything outside this class loader should not be aware that classes are shaded, all classes are referred with their original name.